src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -1058,27 +1246,44 @@ public void ProcessFileChange(string fullFilePath)
                         // all adds/removals of documents for this project happen under our lock, it's safe to do this
                         // check without taking the main workspace lock
                         var document = _project._workspace.CurrentSolution.GetDocument(documentId);
-
                         if (document == null)
                         {
                             return;
                         }
 
+                        _project._documentIdToDynamicFileInfoProvider.TryGetValue(documentId, out var fileInfoProvider);
+
                         _project._workspace.ApplyChangeToWorkspace(w =>
                         {
                             if (w.IsDocumentOpen(documentId))
                             {
                                 return;
                             }
 
-                            var documentInfo =
-                                DocumentInfo.Create(
-                                    document.Id,
-                                    document.Name,
-                                    document.Folders,
-                                    document.SourceCodeKind,
-                                    new FileTextLoader(fullFilePath, defaultEncoding: null),
-                                    document.FilePath);
+                            TextLoader textLoader;
+                            IDocumentServiceProvider documentServiceProvider;
+                            if (fileInfoProvider == null)
+                            {
+                                textLoader = new FileTextLoader(fullFilePath, defaultEncoding: null);
+                                documentServiceProvider = null;
+                            }
+                            else
+                            {
+                                var fileInfo = _project._threadingContext.JoinableTaskFactory.Run(() => fileInfoProvider.GetDynamicFileInfoAsync(_project.Id, _project._filePath, fullFilePath, CancellationToken.None));
+
+                                textLoader = fileInfo.TextLoader;
```

Comment:
To add another datapoint here about JTF - this means that each time we update a document, it has to transition to the fg thread and then back. When testing this on a larger solution this is adding some noticable significant overhead;

Reply:
I just looked JTF.Run source and i dont see anything about switching to UI thread?

just to make sure let me ask @AArnott 

Andrew, if I use JTF.Run(async () => await ...), I get that it will block current thread, but will it switch to UI thread?

I thought JTF.Run will try to run the async method in line if possible but otherwise, run in BG while blocking current thread in a way that if it is called from UI thread, don't block other JTF from running?

Reply:
JTF.Run will not switch to the UI thread without an explicit switch. However, if we're on a background thread and if GetDynamicFileInfoAsync or something further done does switch that's where we run into trouble.

Reply:
@davkean but that's a problem of using async in a synchronous path and not directly related to JTF. correctly used JTF actually help us from getting into dead lock in those cases.

anyway, this is what @jasonmalinowski asked me to do until methods in IWorkspaceProjectContext is changed to Async and whole call path become async path.

Reply:
Yes exactly, sorry - I just had a mental model of how this was implemented under the covers and it didn't match that. We can ignore this whole thread until we flow through async. At that point, I'll call from background thread.

Reply:
@rynowak @heejaechang  `JTF.Run` does *not* transition to the main thread. It invokes the delegate on the caller's thread. The delegate passed to it *may* switch to the main thread.

Reply:
@AArnott - thanks for clearing it up. We were seeing some really wierd behavior whenever I pushed an update through this API. We debugged it an found the real cuprit 👍